stan/math/opencl/kernel_generator/calc_if.hpp

```diff
@@ -0,0 +1,122 @@
+ * @tparam T_a type of first argument
+ * @tparam T_b type of second argument
+ */
+template <bool Do_Calculate, typename T>
+class calc_if_
+    : public operation_cl<calc_if_<Do_Calculate, T>,
+                          typename std::remove_reference_t<T>::Scalar, T> {
+ public:
+  using Scalar = typename std::remove_reference_t<T>::Scalar;
+  using base = operation_cl<calc_if_<Do_Calculate, T>, Scalar, T>;
+  using base::var_name;
+
+ protected:
+  using base::arguments_;
+
+ public:
+  /**
+   * Constructor
+   * @param a expression to calc_if
+   */
+  explicit calc_if_(T&& a) : base(std::forward<T>(a)) {}
+
+  inline kernel_parts generate(const std::string& i, const std::string& j,
+                               const std::string& var_name_arg) const {
+    if (Do_Calculate) {
+      var_name = var_name_arg;
+    }
+    return {};
+  }
+
+  /**
+   * Generates kernel code for assigning this expression into result expression.
+   * @param[in,out] generated set of (pointer to) already generated operations
+   * @param name_gen name generator for this kernel
+   * @param i row index variable name
+   * @param j column index variable name
+   * @param result expression into which result is to be assigned
+   * @return part of kernel with code for this and nested expressions
+   * @throws std::invalid_argument dimensions of expression and result can not
+   * be resized.
+   */
+  template <typename T_result>
+  kernel_parts get_whole_kernel_parts(
+      std::set<const operation_cl_base*>& generated, name_generator& ng,
+      const std::string& i, const std::string& j,
+      const T_result& result) const {
+    if (Do_Calculate) {
+      return std::get<0>(arguments_)
+          .get_whole_kernel_parts(generated, ng, i, j, result);
+    } else {
+      return {};
+    }
+  }
+
+  /**
+   * Sets kernel arguments for nested expressions.
+   * @param[in,out] generated set of expressions that already set their kernel
+   * arguments
+   * @param kernel kernel to set arguments on
+   * @param[in,out] arg_num consecutive number of the first argument to set.
+   * This is incremented for each argument set by this function.
+   */
+  inline void set_args(std::set<const operation_cl_base*>& generated,
+                       cl::Kernel& kernel, int& arg_num) const {
+    if (Do_Calculate) {
+      std::get<0>(arguments_).set_args(generated, kernel, arg_num);
+    }
+  }
+
+  /**
+   * View of a matrix that would be the result of evaluating this expression.
+   * @return view
+   */
+  inline matrix_cl_view view() const { return std::get<0>(arguments_).view(); }
+};
+
+template <bool Do_Calculate, typename T,
+          typename = require_all_valid_expressions_and_none_scalar_t<T>>
+inline calc_if_<Do_Calculate, as_operation_cl_t<T>> calc_if(T&& a) {
+  return calc_if_<Do_Calculate, as_operation_cl_t<T>>(
+      as_operation_cl(std::forward<T>(a)));
+}
+
+namespace internal {
+template <typename T>
+struct is_no_output_impl : std::false_type {};
```

Comment:
Not sure I like the name is_no_output, but dont have a better suggestion. is_without_output? has_no_output (we tend to use is_* so this is probably bad). is_no_ouput_kernel?

Reply:
is_without_output makes most sense to me.